### PR TITLE
Import kubeform - a Kubernetes Terraform plugin.

### PR DESCRIPTION
**Pull request by madhusudancs**
This should probably be in a separate repository, but upon @bgrant0607's suggestion I am putting it here until we find a better home.

Supersedes #1365

### CODE REVIEW

**Comment by mikedanese** [comment]
cc @errordeveloper @colemickens

**Comment by mikedanese** [comment]
We should publish cross builds to kubernetes release bucket if we want to start using this in e.g. kubernetes-anywhere.

File: kubeform/cmd/terraform-provider-kubernetes/terraform.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"github.com/hashicorp/terraform/plugin"
+
+	"k8s.io/kubeform/pkg/provider"
```
Comment (mikedanese) [inline kubeform/cmd/terraform-provider-kubernetes/terraform.go]: Shouldn't this import be k8s.io/contib/kubeform for now?

File: kubeform/pkg/provider/provider.go [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+	d.SetId("")
+
+	return nil
+}
+
+func ReadCluster(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+
+func poll(pollInterval, pollTimeout time.Duration, cond func() (bool, error)) bool {
```
Comment (mikedanese) [inline kubeform/pkg/provider/provider.go]: you could use [wait.Poll](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/wait/wait.go#L163)

**Comment by mikedanese** [comment]
```
!!! 'gofmt -s' needs to be run on the following files: 
./kubeform/pkg/provider/provider.go
./kubeform/pkg/provider/provider_test.go
!!! 'golint' problems: 
/home/travis/gopath/src/k8s.io/contrib/kubeform/pkg/provider/provider.go:35:2: exported const UserAgent should have comment (or a comment on this block) or be unexported
/home/travis/gopath/src/k8s.io/contrib/kubeform/pkg/provider/provider.go:46:1: exported function Provider should have comment or be unexported
/home/travis/gopath/src/k8s.io/contrib/kubeform/pkg/provider/provider.go:134:1: exported function CreateKubeconfig should have comment or be unexported
/home/travis/gopath/src/k8s.io/contrib/kubeform/pkg/provider/provider.go:167:1: exported function DeleteKubeconfig should have comment or be unexported
/home/travis/gopath/src/k8s.io/contrib/kubeform/pkg/provider/provider.go:172:1: exported function ReadKubeconfig should have comment or be unexported
/home/travis/gopath/src/k8s.io/contrib/kubeform/pkg/provider/provider.go:199:1: exported function CreateCluster should have comment or be unexported
/home/travis/gopath/src/k8s.io/contrib/kubeform/pkg/provider/provider.go:217:1: exported function DeleteCluster should have comment or be unexported
/home/travis/gopath/src/k8s.io/contrib/kubeform/pkg/provider/provider.go:240:1: exported function ReadCluster should have comment or be unexported
/home/travis/gopath/src/k8s.io/contrib/kubeform/pkg/provider/provider_test.go:121:6: don't use underscores in Go names; func testCluster_basic should be testClusterBasic

**Comment by mikedanese** [comment]
CI is failing.

**Comment by errordeveloper** [comment]
Looks awesome, we'd love to see a Kubernetes provider for Terraform. From looking at the code, it looks like currently this doesn't implement any of the resources, is it correct or am I missing something? I've seen at least one attempt (https://github.com/hashicorp/terraform/pull/3453), and from my understanding it got stalled on due to missing some features in Terraform, as show in https://github.com/hashicorp/terraform/pull/3453#issuecomment-224261369.

File: kubeform/cmd/terraform-provider-kubernetes/terraform.go (hunk 1 above)
Comment (madhusudancs) [inline kubeform/cmd/terraform-provider-kubernetes/terraform.go]: Can I do this in a separate PR? Here is the PR that fixes the import and adds a Makefile - #1416. I would like to keep this PR just for the initial import of the source files without modification so I can easily keep track of things (running diff etc.). The diffs here already look huge.

File: kubeform/pkg/provider/provider.go (hunk 2 above)
Comment (madhusudancs) [inline kubeform/pkg/provider/provider.go]: Intentionally trying to use as little Kubernetes code as possible as a library. The dependency is ridiculous already. I am building a 60MB binary now. Whenever the external versioned client libraries are available I would like to move to that.

**Comment by madhusudancs** [comment]
> We should publish cross builds to kubernetes release bucket if we want to start using this in e.g. kubernetes-anywhere.

@mikedanese Agreed. Will do.

Fixed the gofmt and golint errors.

PTAL.

**Comment by madhusudancs** [comment]
@errordeveloper this is much more basic than the one you have linked to. This doesn't implement any of the Kubernetes resources. I want to start with something minimal that unblocks people from adopting kubernetes-anywhere. Two of the biggest blockers for kubernetes-anywhere adoption at least as far as I can see are:

1. `kubeconfig` creation/deletion
2. Cleanup of cloud resources that different controllers create. Example, routes created by routecontroller.

I want to start with something that works, in order to demonstrate that this is useful. And then we can collectively add the necessary Kubernetes resources/features on top of that. So please feel free to start sending PRs. I will continue to do so too, where I see fit.

**Pull request merged**